File: app/src/main/java/com/lelloman/demo/fragment/IdenticonFragment.java
package com.lelloman.demo.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewCompat;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.lelloman.identicon.IdenticonDrawable;
import com.lelloman.identicon.IdenticonView;

import java.util.Random;

public abstract class IdenticonFragment extends Fragment{

	public interface IdenticonFragmentListener {
		void onIdenticonSelected(int hash, ImageView identiconView, IdenticonFragment fragment);
	}

	protected static final Random RANDOM = new Random();

	private IdenticonFragmentListener mListener;

	@Override
	public void onAttach(Context context) {
		super.onAttach(context);
		mListener = (IdenticonFragmentListener) context;
	}

	@Override
	public void onDetach() {
		super.onDetach();
		mListener = null;
	}

	@Nullable
	@Override
	public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {

		RecyclerView recyclerView = new RecyclerView(getContext()){
			@Override
			protected void onSizeChanged(int w, int h, int oldw, int oldh) {
				super.onSizeChanged(w, h, oldw, oldh);
				if(w != 0) {
					int columns = (int) (w / (getResources().getDisplayMetrics().density * 64));
					setLayoutManager(new GridLayoutManager(getContext(), columns, GridLayoutManager.VERTICAL, false));
					getLayoutManager().scrollToPosition(Integer.MAX_VALUE / 2);
				}
			}
		};

		recyclerView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
		recyclerView.setAdapter(getAdapter());


		return recyclerView;
	}

	protected abstract RecyclerViewAdapter getAdapter();

	public abstract class RecyclerViewAdapter extends RecyclerView.Adapter<ViewHolder> {

		@Override
		public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
			IdenticonView view = new IdenticonView(getContext()){
				@Override
				protected IdenticonDrawable makeIdenticonDrawable(int width, int height, int hash) {
					return RecyclerViewAdapter.this.makeIdenticonDrawable(width,height,hash);
				}

				@Override
				protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
					super.onMeasure(widthMeasureSpec,heightMeasureSpec);
					setMeasuredDimension(getMeasuredWidth(), getMeasuredWidth());
				}
			};
			view.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

			return new ViewHolder(view);
		}

		@Override
		public void onBindViewHolder(ViewHolder holder, int position) {
			int hash = getHashForPosition(position);
			holder.hash = hash;
			holder.identiconView.setHash(hash);
			ViewCompat.setTransitionName(holder.identiconView, String.valueOf(hash));
		}

		@Override
		public int getItemCount() {
			return Integer.MAX_VALUE;
		}

		protected abstract int getHashForPosition(int position);
		protected abstract IdenticonDrawable makeIdenticonDrawable(int width, int height, int hash);
	}

	public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

		public final IdenticonView identiconView;
		public int hash;

		public ViewHolder(View itemView) {
			super(itemView);
			identiconView = (IdenticonView) itemView;
			itemView.setOnClickListener(this);
		}

		@Override
		public void onClick(View view) {
			identiconView.setTransitionName(String.valueOf(hash));
			if(mListener != null)
				mListener.onIdenticonSelected(hash, identiconView, IdenticonFragment.this);
		}
	}
}

File: app/src/main/java/com/lelloman/demo/fragment/IdenticonFragment.kt
package com.lelloman.demo.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.view.ViewCompat
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import com.lelloman.identicon.drawable.IdenticonDrawable
import com.lelloman.identicon.view.IdenticonView
import java.util.*

abstract class IdenticonFragment : Fragment() {

    private var listener: IdenticonFragmentListener? = null

    protected abstract val adapter: RecyclerViewAdapter

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        listener = context as IdenticonFragmentListener?
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {

        val recyclerView = object : RecyclerView(context!!) {
            override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
                super.onSizeChanged(w, h, oldw, oldh)
                if (w != 0) {
                    val columns = (w / (resources.displayMetrics.density * 64)).toInt()
                    layoutManager = GridLayoutManager(context, columns, GridLayoutManager.VERTICAL, false).apply {
                        scrollToPosition(Integer.MAX_VALUE / 2)
                    }
                }
            }
        }

        recyclerView.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)
        recyclerView.adapter = adapter


        return recyclerView
    }

    companion object {
        val RANDOM = Random()
    }

    abstract inner class RecyclerViewAdapter : RecyclerView.Adapter<ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val view = object : IdenticonView(parent.context) {
                override fun makeIdenticonDrawable(width: Int, height: Int, hash: Int) =
                    this@RecyclerViewAdapter.makeIdenticonDrawable(width, height, hash)

                override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
                    super.onMeasure(widthMeasureSpec, heightMeasureSpec)
                    setMeasuredDimension(measuredWidth, measuredWidth)
                }
            }
            view.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            return ViewHolder(view)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val hash = getHashForPosition(position)
            holder.hash = hash
            holder.identiconView.hash = hash
            ViewCompat.setTransitionName(holder.identiconView, hash.toString())
        }

        override fun getItemCount() = Integer.MAX_VALUE

        protected abstract fun getHashForPosition(position: Int): Int
        protected abstract fun makeIdenticonDrawable(width: Int, height: Int, hash: Int): IdenticonDrawable
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        val identiconView: IdenticonView = itemView as IdenticonView
        var hash: Int = 0

        init {
            itemView.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            identiconView.transitionName = hash.toString()
            listener?.onIdenticonSelected(hash, identiconView, this@IdenticonFragment)
        }
    }

    interface IdenticonFragmentListener {
        fun onIdenticonSelected(hash: Int, identiconView: ImageView, fragment: IdenticonFragment)
    }
}
